cpp/src/arrow/util/async_generator.h

```diff
@@ -1332,4 +1332,49 @@ Result<Iterator<T>> MakeReadaheadIterator(Iterator<T> it, int readahead_queue_si
   return MakeGeneratorIterator(std::move(owned_bg_generator));
 }
 
+/// \brief Make a generator that returns a single pre-generated future
```

Comment:
Will do.

Reply:
Hmm, to make the generator async-reentrant, I think I need to make the future move atomic. That probably implies adding a lock. What do you think?

Reply:
Keep in mind that it is async-reentrant and not sync-reentrant.  The method `operator()` will be completely finished before being called again.  I'm not sure why you would need to make the move atomic?

Reply:
Hmm... does async-reentrant not mean that the generator may be consumed by several threads at once?

Reply:
For example, I see a mutex in MappingGenerator (which doesn't use a background thread).

Reply:
Similarly, VectorGenerator uses an atomic index.

Reply:
The atomic index in VectorGenerator is not needed (mistake on my part).  The mutex in MappingGenerator is neccesary.  Async reentrant means that the `operator()()` method may be called before the future previously returned has completed but after the last call to `operator()()` completed.

So, for example, in MappingGenerator the state is modified by future callbacks.  These future callbacks run after the call to `operator()()` completed so it's possible they will be called at the same time as the next call to `operator()()`.

Reply:
I see, thank you.

Reply:
FWIW: There are only two spots today where we pull in an async-reentrant fashion.  The first is in `ReadaheadGenerator::operator()()`. There is a loop...

```
      for (int i = 0; i < max_readahead_; i++) {
        auto next = source_generator_();
        next.AddCallback(mark_finished_if_done_);
        readahead_queue_.push(std::move(next));
      }
```

If `source_generator_` is fully synchronous (i.e. it always returns finished futures) this does not add any parallelism (an unfortunate fact I hope to remedy someday).

However, if `source_generator_` returns an unfinished future, then this will fan out the tasks.  For example, if `source_generator_` is reading from a file then this will cause up to `max_readahead_` concurrent file reads.

The second spot is in `MergedGenerator` and I will probably remove it at some point as it isn't important and causes more headaches than it should.